### ISSUE TITLE
Debug failing test in Chrome Beta/Canary

### ISSUE THREAD

**Issue by jeffposnick** [GoogleChrome/sw-helpers#242]
**Library Affected**:
`sw-routing`

One of our tests is failing in non-production versions of Chrome:

```sh
[Failed] End to End Test of RegExpRoute > should work properly when there are multiple routes
    expected 'Cannot GET /static\n' to equal 'static response'
    [Stack Trace]
    AssertionError: expected 'Cannot GET /static\n' to equal 'static response'
        at Promise.all.iframe.contentWindow.fetch.then.then (end-to-end.js:7:43)
        at <anonymous>
```

**Comment by gauntface** [comment]
@jeffposnick what does "non-production versions of Chrome" mean?

**Comment by jeffposnick** [comment]
Beta + Dev/Canary.

**Comment by jeffposnick** [comment]
tl;dr it looks like there's a race condition in the `goog.swUtils.activateSW()` promise that, the first time it's called, causes it to resolve before the service worker has taken control of the iframe. @gauntface, this might be something for you to dig into further.

Here's the full test that's failing in Chrome M57+, but works fine in M56. It's from https://github.com/GoogleChrome/sw-helpers/blob/master/packages/sw-routing/test/browser/end-to-end.js:

```
describe('End to End Test of RegExpRoute', function() {
  it('should work properly when there are multiple routes', function() {
    return goog.swUtils.activateSW('../static/regexp-route.js')
      .then((iframe) => Promise.all([
        iframe.contentWindow.fetch('/static')
          .then((response) => response.text())
          .then((text) => expect(text).to.equal('static response')),

        iframe.contentWindow.fetch('/echo3/1st/abc/2nd/def/3rd/ghi')
          .then((response) => response.json())
          .then((json) => expect(json).to.eql(['abc', 'def', 'ghi'])),

        iframe.contentWindow.fetch('/will-not-match')
          .then((response) => expect(response.status).to.equal(404)),
      ]));
  });
});
```

Some observations:
- Those tests all rely on `goog.swUtils.activateSW()` to register and then wait for the SW to activate, at which point they make requests that test the SW's functionality.
- If I reorder the tests in that file, which ever one I list first will fail. E.g. moving `'End to End Test of Router'` to the first position will cause that to fail. The rest of the tests will all succeed, so the failure isn't due to any specific test.
- Forcing a timeout of even 1ms in between the `goog.swUtils.activateSW()` and the calls to `iframe.contentWindow.fetch()` will get the test listed first working:

```js
return goog.swUtils.activateSW('../static/regexp-route.js')
  .then((iframe) => new Promise((resolve) => window.setTimeout(() => resolve(iframe), 1)))
  .then((iframe) => Promise.all([/*...*/]);
```

- I can verify in the Network DevTools panel that when a test fails, the corresponding `fetch()` doesn't get picked up by the service worker, i.e. the service worker hasn't taken control of the iframe:
![screen shot 2017-03-01 at 4 37 01 pm](https://cloud.githubusercontent.com/assets/1749548/23482541/6e17968c-fe9d-11e6-938d-a0f5715e816d.png)

**Comment by gauntface** [comment]
Fixed in latest PR that was merged (hopefully) and opened an issue to add a waitForController method.

**Comment by jeffposnick** [comment]
Thanks for getting the tests green again! Waiting for control is definitely the way to go, and I should have realized from the `activateSW()` name that it was waiting on activation, not control.

Once we have a way to wait on control via the helper library, I'd like to revert the changes in https://github.com/GoogleChrome/sw-helpers/pull/248/commits/63acf8976aed8a7da763bcb68c88f4aa245b78b6. None of those changes should be necessary, though they appear to as a side effect introduced enough of a delay to work around the race condition we were bumping into.

In particular, I also used to write `event.waitUntil(self.clients.claim())` everyone, until it was [pointed out](https://github.com/w3c/ServiceWorker/issues/799#issuecomment-166295313) that the promise returned by `clients.claim()` just resolves immediately, independent of the service worker takes control of any pages. (See [this section of the spec](https://w3c.github.io/ServiceWorker/#clients-claim) for more details.) The same goes for [`skipWaiting()`](https://w3c.github.io/ServiceWorker/#dom-serviceworkerglobalscope-skipwaiting).

**Comment by gauntface** [comment]
Happy to revert once the waitForController method is in sw-testing-helpers.
